### ISSUE TITLE
Pkg.clone shows password in cleartext on screen when cloning via https

### ISSUE THREAD

**Issue by davidanthoff** [JuliaLang/julia#17422]
When I ``Pkg.clone`` a private repo on github using the https URL, I get a prompt that asks for the username, and then a prompt that asks for the password. When I type my password, I can see it in cleartext on the screen. Instead, ``*`` should be shown on the screen for each character I type.

I think this should get the labels ``bug`` and assigned to the 0.5.0 milestone because this is quite a no-no from a security point of view.

@tkelman (Tony, feel free to stop me from CCing you on the things that I think should be assigned to 0.5.0, but I'm treating you as the release manager until I hear push back from you :)

````
julia> versioninfo()
Julia Version 0.5.0-dev+5435
Commit 894fbe7* (2016-07-14 16:00 UTC)
Platform Info:
  System: NT (x86_64-w64-mingw32)
  CPU: Intel(R) Core(TM) i7-4770 CPU @ 3.40GHz
  WORD_SIZE: 64
  BLAS: libopenblas (USE64BITINT DYNAMIC_ARCH NO_AFFINITY Haswell)
  LAPACK: libopenblas64_
  LIBM: libopenlibm
  LLVM: libLLVM-3.7.1 (ORCJIT, haswell)
````

**Comment by tkelman** [comment]
It was either @wildart or @stevengj who wrote the authentication code I believe.

Re: pinging me personally, I'd rather you didn't, x-ref https://github.com/JuliaLang/julia/issues/15068#issuecomment-232776978

**Comment by tkelman** [comment]
I can reproduce this. @wildart what's the best way to find out where in the call chain libgit2 gets to for cloning a private https package?

**Comment by tkelman** [comment]
I'm very confused. When I call `LibGit2.prompt` with `password=true` by itself, it works as it should. But not within `Pkg.clone` which should be calling https://github.com/JuliaLang/julia/blob/edb112af0b058d1f5a815fc6cb524f586cd18bdd/base/libgit2/callbacks.jl#L175
```
julia> a = LibGit2.prompt("foo", password=true);
foo:
julia> a
"blah"
```

**Comment by wildart** [comment]
ha, it's a terminal issue. `Base.getpass` is a culprit.

**Comment by tkelman** [comment]
Happy to test if you have a proposed patch/PR.

**Comment by stevengj** [comment]
For `getpass` on Windows, see #8228.  We ended up using the implementation based on [_getch](https://msdn.microsoft.com/en-us/library/078sfkak.aspx), which is *supposed* to read a character without echoing it on the terminal.

**Comment by wildart** [comment]
I can confirm that `Base.getpass` on windows, shows doesn't mask input and hangs causing #17423

**Comment by tkelman** [comment]
It looks like it works differently depending on whether you're running from mintty (in cygwin or msys2) or cmd.

**Comment by wildart** [comment]
I tried 0.4.5 and `ccall(:_getch, UInt8, ())` works fine. On 0.5 cygwin build this call hangs.

**Comment by tkelman** [comment]
It's not a difference between 0.4 and 0.5, it's a difference in what terminal you're using to Julia. `_getch` doesn't know how to deal with the named pipes that cygwin's mintty terminal emulator uses. We might need to use something smarter if we want this to work properly in mintty. But note that this is a separate issue from the libgit2 authentication issue, since that happens even outside of mintty. As I said above, `LibGit2.prompt` and `Base.getpass` work fine from cmd, but `Pkg.clone` of a private package doesn't.

**Comment by wildart** [comment]
Question is from where `Pkg.clone` was executed? If from `mintty` terminal then it would hang because of `Base.getpass`.

**Comment by tkelman** [comment]
> Question is from where `Pkg.clone` was executed?

cmd in my tests. Probably powershell for @davidanthoff, but that usually uses the same console host.

**Comment by wildart** [comment]
I'm running julia from cmd and it works fine.

**Comment by tkelman** [comment]
You can clone a private package over https with recent 0.5-dev in cmd, without the password getting echoed? I seem to get the password echoed sometimes in cmd if Pkg.clone is the first thing I run, but not if I do anything else in the REPL first.

**Comment by wildart** [comment]
Yes, no echo for me. The only problem that enter is swallowed and following output starts right after password prompt.

**Comment by davidanthoff** [comment]
I'm not using mintty, I'm just starting julia from the startmenu shortcut to reproduce this, so there is also no powershell involved.

Things got even weirder just now. I essentially got this output (I added line numbers in ``[]`` that are not present in the original output, and replaced my password with ``mypassword``):
````
   _       _ _(_)_     |  A fresh approach to technical computing
  (_)     | (_) (_)    |  Documentation: http://docs.julialang.org
   _ _   _| |_  __ _   |  Type "?help" for help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 0.5.0-dev+5478 (2016-07-18 02:34 UTC)
 _/ |\__'_|_|_|\__'_|  |  Commit 43f3c05 (0 days old master)
|__/                   |  x86_64-w64-mingw32

julia> Pkg.clone("https://github.com/davidanthoff/Judyp.jl.git")
[1] INFO: Cloning Judyp from https://github.com/davidanthoff/Judyp.jl.git
[2] Username for 'https://github.com':davidanthoff
[3] Password for 'https://davidanthoff@github.com':mypassword
[4] Username for 'https://github.com':Password for 'https://mypassword@github.com':
````
I typed in everything. In line 2, I entered my username and then hit Enter. In line 3 I entered my password and hit Enter. Then nothing happened. Then I hit Enter again, and line 4 appeared. And that is now echoing back the *password* that I entered as the username in front of the ``@github`` thing.

**Comment by davidanthoff** [comment]
Oh, and I get the same behavior that @tkelman mentioned, i.e. if I execute any simple julia command at the REPL first, things work:

````
   _       _ _(_)_     |  A fresh approach to technical computing
  (_)     | (_) (_)    |  Documentation: http://docs.julialang.org
   _ _   _| |_  __ _   |  Type "?help" for help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 0.5.0-dev+5478 (2016-07-18 02:34 UTC)
 _/ |\__'_|_|_|\__'_|  |  Commit 43f3c05 (0 days old master)
|__/                   |  x86_64-w64-mingw32

julia> 5*5
25

julia> Pkg.clone("https://github.com/davidanthoff/Judyp.jl.git")
INFO: Cloning Judyp from https://github.com/davidanthoff/Judyp.jl.git
Username for 'https://github.com':davidanthoff
Password for 'https://davidanthoff@github.com':INFO: Computing changes...
INFO: No packages to install, update or remove

julia>
````
There should still be a line-break before the ``INFO: Computing changes...``, but no password is echoed.

**Comment by stevengj** [comment]
Note that Python's `getpass` function [similarly calls msvcrt.getwch](https://svn.python.org/projects/python/tags/r32/Lib/getpass.py) on Windows (see `win_getpass`), which calls [_getwch()](https://svn.python.org/projects/python/tags/r32/PC/msvcrtmodule.c) (see `msvcrt_getwch`).   Does Python's `getpass` function work in mintty?

**Comment by stevengj** [comment]
cc @Keno on the terminal weirdness here.

**Comment by tkelman** [comment]
No the native version of Python is pretty much unusable in mintty, which we've generally been doing a much better job of. How hard can it be to not echo input back to the screen, do we need to do a ccall for that?

**Comment by stevengj** [comment]
Most of the sources I can find suggest that you use `tcgetattr`/`tcsetattr` suppress terminal output (see e.g. [here](http://stackoverflow.com/questions/6856635/hide-password-input-on-terminal)).  No idea how that interacts with mintty.  (My suspicion is that `getch_` does this already, so we won't gain anything.)

**Comment by stevengj** [comment]
We can look at how [pdcurses](http://pdcurses.sourceforge.net/) and cygwin's ncurses work.

**Comment by Keno** [comment]
Maybe just pop up a text box? Windows does have that available as an API function if I remember correctly.

**Comment by stevengj** [comment]
(See also mintty/mintty#56)

**Comment by tkelman** [comment]
that might be better if we're prompting for interaction anyway

**Comment by Keno** [comment]
Unfortunately I can't find the API function with text input. There's one with buttons, but unless we want the user to encode their password in ternary first that probably doesn't work.

**Comment by vtjnash** [comment]
> How hard can it be to not echo input back to the screen, do we need to do a ccall for that?

extremely awful. We fork stty.exe twice every time you hit a character at the repl to provide this. (the alternative is to build against cygwin1.dll and hope the user gets lucky with libc dll versioning + fork + rebase triple hell all matching against the buildbot)

If GPL2 is OK for this, there's pinentry, https://www.gnupg.org/download/, implementing the Assuan protocol https://www.gnupg.org/documentation/manuals/assuan/index.html

**Comment by Keno** [comment]
Found something: https://msdn.microsoft.com/en-us/library/windows/desktop/aa375177(v=vs.85).aspx

**Comment by tkelman** [comment]
I don't think GPL is okay for this unless we make a separate wrapper program that we shell out to. winapi looks more promising.

**Comment by Keno** [comment]
A little antiquated perhaps, but it's ok:

![screen shot 2016-07-19 at 7 51 29 pm](https://cloud.githubusercontent.com/assets/1291671/16970467/46cebee8-4dea-11e6-8d82-dfcf8f0014a5.png)

**Comment by tkelman** [comment]
That's probably why it says to use the newer API when you're on Vista or newer?

**Comment by Keno** [comment]
Yes, the new API is super complicated, but I managed to figure it out as well just now (though this is the server version, so the UI is different from what it would be on the desktop version):
![screen shot 2016-07-19 at 8 23 51 pm](https://cloud.githubusercontent.com/assets/1291671/16971081/bc36b786-4dee-11e6-9be3-4e049cdc121f.png)

**Comment by stevengj** [comment]
Some software (e.g. [winpty](https://github.com/rprichard/winpty)) actually just creates a hidden console window for user input.

**Comment by davidanthoff** [comment]
I think any of these dialog boxes would be ok for now, i.e. fix this security bug. Long term it would be nicer if the password prompt was console based, but imo that could come after 0.5.0.

**Comment by davidanthoff** [comment]
I'm closing this, seems resolved.

**Comment by stevengj** [comment]
Yes, fixed by #17506.
